include/unifex/filter_stream.hpp

```diff
@@ -0,0 +1,250 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/bind_back.hpp>
+#include <unifex/manual_lifetime.hpp>
+#include <unifex/std_concepts.hpp>
+#include <unifex/stream_concepts.hpp>
+
+#include <exception>
+#include <functional>
+#include <type_traits>
+#include <utility>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+namespace _filter {
+template <typename StreamSender, typename FilterFunc, typename Receiver>
+struct _op {
+  struct type;
+};
+template <typename StreamSender, typename FilterFunc, typename Receiver>
+using operation = typename _op<
+    remove_cvref_t<StreamSender>,
+    remove_cvref_t<FilterFunc>,
+    remove_cvref_t<Receiver>>::type;
+
+template <typename StreamSender, typename FilterFunc, typename Receiver>
+struct _next_receiver {
+  struct type;
+};
+template <typename StreamSender, typename FilterFunc, typename Receiver>
+using next_receiver = typename _next_receiver<
+    remove_cvref_t<StreamSender>,
+    remove_cvref_t<FilterFunc>,
+    remove_cvref_t<Receiver>>::type;
+
+template <typename StreamSender, typename FilterFunc, typename Receiver>
+struct _next_receiver<StreamSender, FilterFunc, Receiver>::type {
+  using next_receiver_t = next_receiver<StreamSender, FilterFunc, Receiver>;
+  operation<StreamSender, FilterFunc, Receiver>& op_;
+
+  template(typename CPO)                       //
+      (requires is_receiver_query_cpo_v<CPO>)  //
+      friend auto tag_invoke(CPO cpo, const type& r) noexcept(
+          is_nothrow_callable_v<CPO, const Receiver&>)
+          -> callable_result_t<CPO, const Receiver&> {
+    return std::move(cpo)(std::as_const(r.op_.receiver_));
+  }
+
+#if UNIFEX_ENABLE_CONTINUATION_VISITATIONS
+  template <typename Func>
+  friend void
+  tag_invoke(tag_t<visit_continuations>, const type& r, Func&& func) {
+    std::invoke(func, r.op_.receiver_);
+  }
+#endif
+
+  template <typename... Values>
+  void set_value(Values... values) && noexcept {
+    auto& op = op_;
+    op.next_.destruct();
+    UNIFEX_TRY {
+      const bool doFilter =
+          !std::invoke(op.filter_, std::forward<Values>(values)...);
+
+      if (doFilter) {
+        op.next_.construct_with([&] {
+          return unifex::connect(next(op.stream_), next_receiver_t{op});
+        });
+        unifex::start(op.next_.get());
+      } else {
+        unifex::set_value(std::move(op.receiver_), std::move(values)...);
+      }
+    }
+    UNIFEX_CATCH(...) {
+      unifex::set_error(std::move(op.receiver_), std::current_exception());
+    }
```

Comment:
If we rely on the operation state's destructor to destroy `next_`, the lifetime of `values...` as forwarding references can be made safe (i.e. we can forward the values into `set_value` even if they happen to refer to values stored somewhere in `next_`).

Unfortunately, I've realized while typing this out that, because the call to `connect` might throw, we need a `bool` to keep track of whether `next_` contains a value to be destructed in `op`'s destructor. :(

```suggestion
    UNIFEX_TRY {
      const bool doFilter =
          !std::invoke(op.filter_, std::as_const(values)...);

      if (doFilter) {
        op.next_.destruct();
        op.nextEngaged_ = false;
        op.next_.construct_with([&] {
          return unifex::connect(next(op.stream_), next_receiver_t{op});
        });
        op.nextEngaged_ = true;
        unifex::start(op.next_.get());
      } else {
        unifex::set_value(std::move(op.receiver_), std::forward<Values>(values)...);
      }
    }
    UNIFEX_CATCH(...) {
      unifex::set_error(std::move(op.receiver_), std::current_exception());
    }
```